Migrate 'Same-origin' from Sanity to MDX (#289)
clerkinc/clerk-docs: docs/backend-requests/making/same-origin.mdx

---
title: Same-origin requests

> description: Learn about various ways to make same-origin requests to the backend using the Clerk SDK.

Reviewer: Okay it looks like we use `Clerk SDK` quite a bit. It sounds kind of awkward to me though...what do you think is better?

* Clerk SDK
* Clerk's SDK
* Clerk SDKs
* Clerk's SDKs

Author: We could cut SDK altogether and just say "Learn about various ways to make same-origin requests to the backend." or "Learn about various ways to make same-origin requests to the backend when using Clerk SDKs."

Reviewer: I like "Clerk SDKs" 👍

<CodeBlockTabs options={["index.jsx", "index.tsx"]}>
```jsx
import useSWR from 'swr';
import { useAuth } from '@clerk/nextjs';

export default function useClerkSWR(url) {
  const { getToken } = useAuth();

  const fetcher = async (...args) => {
    return fetch(...args, {
      headers: { Authorization: `Bearer ${await getToken()}` }
    }).then(res => res.json());
  };

  return useSWR(url, fetcher);
}
```

```tsx
import useSWR from 'swr';
import { useAuth } from '@clerk/nextjs';

export default function useClerkSWR(url: string) {
  const { getToken } = useAuth();

  const fetcher = async (...args: [RequestInfo]) => {
    return fetch(...args, {
      headers: { Authorization: `Bearer ${await getToken()}` }
    }).then(res => res.json());
  };

  return useSWR(url, fetcher);
}
```
</CodeBlockTabs>

When you re-focus a page or switch between tabs, SWR automatically revalidates data. This can be useful to immediately synchronize to the latest state in scenarios like stale mobile tabs, or laptops that went to sleep.

> ## Tanstack Query (FKA React Query)

Reviewer: ```suggestion
## Tanstack Query (React Query)
```

> If you are using [Tanstack Query](https://tanstack.com/query/latest), formally known as React Query, in your application, you need to use a query function that throws errors. Since the native Fetch API does not, you can add it in.

Reviewer: Based on the other similar page:

```suggestion
If you are using [Tanstack Query](https://tanstack.com/query/v4/docs/react/overview), formally known as React Query, in your application, you need to use a query function that throws errors. Since the native Fetch API does not, you can add it in.
```

Reviewer: Should we align this with the other page in general? Let's also take Lennart's feedback into consideration here 🙏

<Callout type="warning">
  Make sure that you have your application wrapped in `<QueryClientProvider />` with a `QueryClient` set.
</Callout>

```jsx
import { useQuery } from '@tanstack/react-query';

const { data, error } = useQuery('foo', async () => {
  const res = await fetch('/api/foo');
  if (!res.ok) {
    throw new Error('Network response error')
  }
  return res.json()
});
```

## Background Fetching

> For applications that are fetching content in the background, when a tab is no longer focused, you will need to include an Authorization header along with your request.

Reviewer: ```suggestion
For applications that are fetching content in the background, like when a tab is no longer focused, you will need to include an Authorization header along with your request.
```
